src/lib.rs

```diff
@@ -619,34 +619,175 @@ impl Bump {
     where
         F: FnOnce() -> T,
     {
-        #[inline(always)]
-        unsafe fn inner_writer<T, F>(ptr: *mut T, f: F)
-        where
-            F: FnOnce() -> T,
-        {
-            // This function is translated as:
-            // - allocate space for a T on the stack
-            // - call f() with the return value being put onto this stack space
-            // - memcpy from the stack to the heap
-            //
-            // Ideally we want LLVM to always realize that doing a stack
-            // allocation is unnecessary and optimize the code so it writes
-            // directly into the heap instead. It seems we get it to realize
-            // this most consistently if we put this critical line into it's
-            // own function instead of inlining it into the surrounding code.
-            ptr::write(ptr, f())
-        }
-
         let layout = Layout::new::<T>();
 
         unsafe {
             let p = self.alloc_layout(layout);
             let p = p.as_ptr() as *mut T;
-            inner_writer(p, f);
+            alloc_with_writer(p, f);
             &mut *p
         }
     }
 
+    /// Tries to pre-allocate space for an object in this `Bump`, initializes
+    /// it using the closure, then returns an exclusive reference to it.
+    ///
+    //TODO: Fix or remove the example here. Should there be a try_alloc method?
+    /// Calling `bump.alloc(x)` is essentially equivalent to calling
+    /// `bump.try_alloc_with(|| x)`. However if you use `try_alloc_with`, then the
+    /// closure will not be invoked until after allocating space for storing
+    /// `x` on the heap.
```

Comment:
As in the TODO comment, should I add a `try_alloc` method too? I think it would nicely fill a

|                     | fallible allocation  | infallible allocation |
|---------------------|----------------------|-----------------------|
| **by value**        | `try_alloc`          | `alloc`               |
| **infallible init** | `try_alloc_with`     | `alloc_with`          |
| **fallible init**   | `try_alloc_try_with` | `alloc_try_with`      |

pattern, but I'd also understand not wanting to promote the by-value version.

Reply:
I added all of them for now, since it was next to no effort. Matching tests will follow later this week.